### ISSUE TITLE
NSRangeException in MLCTrainingGraph on mac intel

### ISSUE THREAD

**Issue by invoxiaehu** [apple/tensorflow_macos#20]
Here's my trace:

2020-11-20 11:35:11.019757: I tensorflow/compiler/mlir/mlir_graph_optimization_pass.cc:116] None of the MLIR optimization passes are enabled (registered 2)
2020-11-20 11:35:11.669 Python[62946:1516017] *** Terminating app due to uncaught exception 'NSRangeException', reason: '*** -[__NSArrayM objectAtIndexedSubscript:]: index 2 beyond bounds [0 .. 1]'
*** First throw call stack:
(
	0   CoreFoundation                      0x00007fff204856af __exceptionPreprocess + 242
	1   libobjc.A.dylib                     0x00007fff201bd3c9 objc_exception_throw + 48
	2   CoreFoundation                      0x00007fff20539a9a -[__NSCFString characterAtIndex:].cold.1 + 0
	3   CoreFoundation                      0x00007fff203f8e31 -[__NSArrayM objectAtIndexedSubscript:] + 169
	4   MLCompute                           0x00007fff2a03aa9b -[MLCTrainingGraph executeGradientFromLayerIndex:batchSize:] + 397
	5   MLCompute                           0x00007fff2a03f547 -[MLCTrainingGraph executeGradientWithBatchSize:options:outputsData:completionHandler:] + 1000
	6   _pywrap_tensorflow_internal.so      0x000000012112f325 _ZN10tensorflow9mlcompute3ops13MLCSubgraphOp23ExecuteMLCTrainingGraphEPNS_15OpKernelContextEPNS1_27MLCSubgraphExecutionContextEj + 533
	7   _pywrap_tensorflow_internal.so      0x000000012112e591 _ZN10tensorflow9mlcompute3ops13MLCSubgraphOp20ProcessMLCSubgraphOpEPNS_15OpKernelContextEPNS1_27MLCSubgraphExecutionContextE + 1205
	8   _pywrap_tensorflow_internal.so      0x000000012113163a _ZN10tensorflow9mlcompute3ops13MLCSubgraphOp7ComputeEPNS_15OpKernelContextE + 1208
	9   libtensorflow_framework.2.dylib     0x0000000135f6821d _ZN10tensorflow12_GLOBAL__N_113ExecutorStateINS_21SimplePropagatorStateEE7ProcessENS2_10TaggedNodeEx + 3765
	10  libtensorflow_framework.2.dylib     0x0000000135fe5f3b _ZN5Eigen15ThreadPoolTemplIN10tensorflow6thread16EigenEnvironmentEE10WorkerLoopEi + 605
	11  libtensorflow_framework.2.dylib     0x0000000135fe5be2 _ZZN10tensorflow6thread16EigenEnvironment12CreateThreadENSt3__18functionIFvvEEEENKUlvE_clEv + 66
	12  libtensorflow_framework.2.dylib     0x0000000135fd6eb1 _ZN10tensorflow12_GLOBAL__N_17PThread8ThreadFnEPv + 97
	13  libsystem_pthread.dylib             0x00007fff20313950 _pthread_start + 224
	14  libsystem_pthread.dylib             0x00007fff2030f47b thread_start + 15
)
libc++abi.dylib: terminating with uncaught exception of type NSException
/usr/local/Cellar/python@3.8/3.8.5/Frameworks/Python.framework/Versions/3.8/lib/python3.8/multiprocessing/resource_tracker.py:216: UserWarning: resource_tracker: There appear to be 3 leaked semaphore objects to clean up at shutdown

**Comment by tux-o-matic** [comment]
It's a more complete stracktrace but the last line is similar to my problem in #19. Was that stack trace printed in the stdout or you looked up for it in the Console @invoxiaehu ?

**Comment by invoxiaehu** [comment]
Hello @tux-o-matic. Thank you for your answer. The trace was in the terminal where I launched jupyter-lab. I work in a notebook in jupyter-lab, with the kernel tensorflow_macos_venv ( with ipykernel)

**Comment by tux-o-matic** [comment]
@invoxiaehu, if you let ML Compute chose the device, you might be running on CPU. In my case, forcing GPU fixes the issue (but might be slower than regular TF on CPU).

**Comment by invoxiaehu** [comment]
@tux-o-matic, When i choose gpu, it is training indeed, but something else is wrong because the loss is nan. Also it is slower the cpu.

**Comment by anna-tikhonova** [comment]
@invoxiaehu Thank you for reporting this issue. Could you, please, point us to or attach an example you are running? This way, we can reproduce this issue locally and investigate.
